### ISSUE TITLE
Events not forwarded to ZeroMQ

### ISSUE THREAD

**Issue by bbkrr** [wazuh/wazuh#11391]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.2.5 | Server | Manager | Sources | Ubuntu 16.04.1 LTS |

I enabled ZeroMQ during compilation of source code using `make TARGET=server USE_ZEROMQ=yes`, and configured for ZeroMQ as:

```
<global>
     ...
    <zeromq_output>yes</zeromq_output>
    <zeromq_uri>ZEROMQ_SOCKET_PATH</zeromq_uri>
</global>
```

Even after doing so, I was not able to get logs in ZeroMQ, am I missing something ? 

Also, if events can be forwarded to ZeroMQ, can we disable alerts.log and not record any events in there and simply forward the events to ZeroMQ ?

**Comment by vikman90** [comment]
Hi @bbkrr,

I have followed your steps. I assume:
```sh
cd src
make clean-internals
make TARGET=server USE_ZEROMQ=yes
cd ..
sudo ./install.sh
```

I confirm that your configuration is valid, and the manager attempts to send the alerts. Can you confirm that this error is not appearing in the log?
```
2021/12/17 17:02:28 wazuh-analysisd: ERROR: Unable to bind the ZeroMQ Socket: /root/zeromq.sock.
```

Regarding your last question, you can disable both plain-text and JSON alerts:
```xml
<jsonout_output>no</jsonout_output>
<alerts_log>no</alerts_log>
```
Analysisd will continue sending the alerts.

Hope that helps.
Best regards.

**Comment by bbkrr** [comment]
Hi @vikman90,

Thank you for the information regarding my last question, it is really helpful.

About the error you mentioned on ZeroMQ, I didn't get such error. There is one error which appears even if I re-install wazuh, but it is not related to ZeroMQ.

```
2021/12/18 03:20:48 wazuh-modulesd:agent-upgrade: INFO: (8153): Module Agent Upgrade started.
2021/12/18 03:20:48 wazuh-modulesd:syscollector: INFO: Module started.
2021/12/18 03:20:48 wazuh-modulesd:syscollector: INFO: Starting evaluation.
2021/12/18 03:20:48 wazuh-modulesd:syscollector: ERROR: [json.exception.out_of_range.403] key 'name' not found
2021/12/18 03:20:48 sca: INFO: Starting evaluation of policy: '/var/ossec/ruleset/sca/cis_debian9.yml'
2021/12/18 03:20:48 wazuh-modulesd:syscollector: INFO: Evaluation finished.
2021/12/18 03:20:51 wazuh-syscheckd: INFO: (6009): File integrity monitoring scan ended.
2021/12/18 03:20:54 sca: INFO: Evaluation finished for policy '/var/ossec/ruleset/sca/cis_debian9.yml'
2021/12/18 03:20:54 sca: INFO: Security Configuration Assessment scan finished. Duration: 6 seconds.
2021/12/18 03:23:03 rootcheck: INFO: Ending rootcheck scan.
```

Also, I get the logs in my `alerts.log` file but I don't see it in my ZeroMQ.

**Comment by Dwordcito** [comment]
Hi @bbkrr I hope you are well

could you tell me the output of "ip address" ?
I suspect that an interface type that you have causes this problem. Are you connected to a VPN in that environment?

**Comment by bbkrr** [comment]
Hi @Dwordcito,

Yes, I am connected to a VPN environment. What role does an interface play here ?

Below is my interface information.

![tmp_93f4c2af-6bee-4867-b88c-d586a8676369](https://user-images.githubusercontent.com/85162559/146739068-e90489e7-1319-436e-85e2-186b6b97b6cb.jpg)

**Comment by bbkrr** [comment]
> Hi @Dwordcito,
> 
> Yes, I am connected to a VPN environment. What role does an interface play here ?
> 
> Below is my interface information.
> 
> ![tmp_93f4c2af-6bee-4867-b88c-d586a8676369](https://user-images.githubusercontent.com/85162559/146739068-e90489e7-1319-436e-85e2-186b6b97b6cb.jpg)

I also tested the same thing in my local VM but got the same issue, i.e. not able to forward logs to zeromq.

**Comment by bbkrr** [comment]
@vikman90 

I was testing ZeroMQ with `zmq.REQ` and `zmq.STREAM`. I tested with `STREAM` because I was not getting any logs from the other method, but even I changed it to STREAM it does not seem to have any difference and I am not receiving any logs in ZeroMQ.

Even though I followed every step from the documentation and as you mentioned in above comment, it is not working as written in the docs, and also there are no error logs to debug the issue.

Do you have any idea, what could be the issue ?

**Comment by vikman90** [comment]
Hi @bbkrr,

Honestly, I hardly know ZeroMQ, but I've made the following test:

- Analysisd implements a publisher ([ZeroMQ Guide - Chapter 1](https://zguide.zeromq.org/docs/chapter1/)).
- We need to implement a subscriber ([ZeroMQ Guide - Chapter 2](https://zguide.zeromq.org/docs/chapter2/)). This is a very basic program I've picked up from the guide:

```c
#include <zmq.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>

int main() {
    void * context = zmq_ctx_new();
    void * subscriber = zmq_socket(context, ZMQ_SUB);
    zmq_connect (subscriber, "tcp://localhost:5555");
    zmq_setsockopt (subscriber, ZMQ_SUBSCRIBE, "", 0);

    while (1) {
        char buffer[4096];
        zmq_recv(subscriber, buffer, sizeof(buffer), 0);
        printf("%s\n", buffer);
    }

    return 0;
}
```

This application subscribes to the manager's alerts by connecting to `localhost:5555`.

```sh
$ gcc -O2 -o zmq-recv zmq-recv.c -lzmq
```

- Manager's configuration:
```xml
<ossec_config>
  <global>
    <zeromq_output>yes</zeromq_output>
    <zeromq_uri>tcp://*:5555</zeromq_uri>
  </global>
</ossec_config>
```

- Start the manager. I saw this error but I suspect that is due to a change in the ZeroMQ API, that works:
```sh
$ /var/ossec/bin/wazuh-control start
2022/01/11 18:37:46 wazuh-analysisd: ERROR: Unable to bind the ZeroMQ Socket: tcp://*:5555.
E: 22-01-11 18:37:47 [21592]dangling 'PUB' socket created at analysisd/output/zeromq.c:61
```

- Run the subscriber:
```sh
$ ./zmq-recv
ossec.alerts
{"timestamp":"2022-01-11T18:44:10.396+0100","rule":{"level":3,"description":"Ossec server started.","id":"502","firedtimes":2,"mail":false,"groups":["ossec"],"pci_dss":["10.6.1"],"gpg13":["10.1"],"gdpr":["IV_35.7.d"],"hipaa":["164.312.b"],"nist_800_53":["AU.6"],"tsc":["CC7.2","CC7.3"]},"agent":{"id":"000","name":"Rocket"},"manager":{"name":"Rocket"},"id":"1641923050.2938542","full_log":"ossec: Ossec started.","decoder":{"name":"ossec"},"location":"wazuh-monitord"}
```

Which matches the same alert logged on `alerts.json`.

I suspect that the ZeroMQ output code may need some maintenance, but I managed to get it working this way. If you keep having troubles, please let us know how you are subscribing to the manager to get the alerts.

Best regards.

**Comment by bbkrr** [comment]
@vikman90 

I subscribed the manager same as you did, but still no log is received.

Are you using same OS and version that I mentioned in the question ?

**Comment by bbkrr** [comment]
@vikman90 

Thank you for your help, it looks like the issue I am facing is already reported earlier and the solution is yet to be released as mentioned in [#8141](https://github.com/wazuh/wazuh/issues/8141) and also in the new issue #11840. I am able to solve some part of the issue referring to these other issues, so we can close this.

Regards
